api/flask/classifier.py

```diff
@@ -1,56 +1,36 @@
-from sklearn.pipeline import Pipeline
-from sklearn.feature_extraction.text import CountVectorizer
-from sklearn.feature_extraction.text import TfidfTransformer
-from sklearn.model_selection import GridSearchCV
-# from sklearn.naive_bayes import MultinomialNB
-from sklearn.linear_model import SGDClassifier
-from collections import Counter
-from sklearn.exceptions import NotFittedError
-
+from threading import Thread
+from multiprocessing import Process
+import sgdclassifier
 import dal
 
-# support vector machine classifier
-text_clf = Pipeline([
-    ('vect', CountVectorizer()),
-    ('tfidf', TfidfTransformer()),
-    ('clf', SGDClassifier(
-        loss='hinge', penalty='l2',
-        alpha=1e-3, random_state=42,
-        max_iter=5, tol=None))])
-
-# exhaustively search for best classifier parameters
-parameters = {
-    'vect__ngram_range': [(1, 1), (1, 2)],
-    'tfidf__use_idf': (True, False),
-    'clf__alpha': (1e-2, 1e-3)}
-gs_clf = GridSearchCV(text_clf, parameters, n_jobs=-1)
-# hack because we can't otherwise tell whether or not gs_clf has been fit yet.
-initialized = False
-
 def fit(uid):
     """
-    Fits a new classifier from the given user's labeled text in the database.
+    Asynchronously spawns a thread which spawns a subprocess which
+    fits a new classifier from the given user's labeled text in the
+    database. We can't just do this in a thread because the fit
+    operation uses multiprocessing and so must be done on the main
+    thread. And we can't just spawn a process because we have to
+    immediately join() it in order to avoid creating zombies. So we
+    spawn a thread whose only responsibility is spawning and joining
+    the process.
+    """
+    t = Thread(target=async_fit, args=(uid,))
+    t.start()
 
-    TODO: persist the given user's new classifier rather than persisting it as
-    the global classifier.
+def async_fit(uid):
+    """
+    Asynchronously spawns a subprocess which fits a new classifier.
     """
-    labeled_text = dal.get_labeled_text(uid)
-    # Requires enough data for each target to be split into training, test, and
-    # validation sets.
-    if len(labeled_text['targets']) >= 2 and all(i >= 3 for i in Counter(labeled_text['targets']).values()):
-        gs_clf.fit(labeled_text['data'], labeled_text['targets'])
-        global initialized
-        initialized = True
+    p = Process(target=sgdclassifier.fit, args=(uid,))
```

Comment:
I assume the comma doesn't matter

Reply:
The comma is intentional and necessary because it's expecting a tuple. But we don't need to pass anything in the second part of the tuple.